adam_preprocessing/shape_stroke_extraction.py

```diff
@@ -108,6 +110,309 @@ def kp2stroke(strokeinfo):
+        """Check if two strokes overlap at their endpoints."""
+        return any(
+            kings_move_distance(stroke1[s1_endpoint], stroke2[s2_endpoint]) == 0
+            for s1_endpoint in [0, -1]
+            for s2_endpoint in [0, -1]
+        )
+
+    def consistent_orientation(stroke1: Sequence[Tuple[float, float]], stroke2: Sequence[Tuple[float, float]]) -> bool:
+        """Check if two overlapping strokes are oriented consistently."""
+        return (kings_move_distance(stroke1[0], stroke2[-1]) == 0) or (kings_move_distance(stroke1[-1], stroke2[0]) == 0)
+
+    # Create a graph where the nodes are strokes and edges represent "overlap at endpoints"
+    g = nx.Graph()
+    g.add_nodes_from(range(len(strokes)))
+    for idx1, stroke1 in enumerate(strokes):
+        for idx2, stroke2 in enumerate(strokes[idx1 + 1 :], start=idx1 + 1):
+            if overlap(stroke1, stroke2):
+                g.add_edge(idx1, idx2)
+
+    if not nx.is_connected(g):
+        raise ValueError("Can't order strokes end to end: Strokes are not connected enough.")
+
+    # The result should be a path graph. Equivalently (https://en.wikipedia.org/wiki/Path_graph):
+    # It should have no cycles (= it's a tree), and no vertex with degree > 2.
+    if nx.is_tree(g) and all(g.degree(n) <= 2 for n in g.nodes):
+        terminals = [n for n in g.nodes if g.degree(n) == 1]
+        last_node = None
+        cur_node = terminals[0]
+        ordering = [terminals[0]]
+        preserve_orientation = [True]
+        while cur_node != terminals[1]:
+            next_node = [n for n in g.neighbors(cur_node) if n != last_node][0]
+            ordering.append(next_node)
+            preserve_orientation.append(consistent_orientation(strokes[cur_node], strokes[next_node]))
+            last_node = cur_node
+            cur_node = next_node
+
+        result = [
+            strokes[stroke_idx] if preserve_orientation else reversed(strokes[stroke_idx])
+            for stroke_idx, preserve_orientation in zip(ordering, preserve_orientation)
+        ]
+        assert len(result) == len(strokes)
+        return result
+    else:
+        raise ValueError("Strokes don't connect in a line, so can't order them end to end.")
+
+
+def plot_oriented_strokes(ax: plt.Axes, strokes: Sequence[Sequence[Tuple[float, float]]]) -> None:
+    """
+    Plot oriented strokes on the given axes.
+
+    This creates a line chart with one color per stroke where each line has arrows pointing in the
+    direction that the given stroke is oriented.
+
+    For strokes with just one point, we plot a circle instead.
+
+    Parameters:
+        ax: The axes to plot on.
+        strokes: The strokes to plot.
+    """
+    from matplotlib.rcsetup import cycler
+    import matplotlib.patheffects as patheffects
+    from itertools import cycle
+    colors = cycler(color=['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf'])
```

Comment:
Nit: I'm assuming these are from a copy/paste operation? No need to move them just to modify this function though.

Reply:
Good point. I sketched these while I was debugging and forgot to move the imports. I'll make the change before merging.

Reply:
Moved these to top level.